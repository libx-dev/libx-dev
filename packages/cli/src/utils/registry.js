/**
 * レジストリ操作ユーティリティ
 *
 * registry/docs.json の読み込み、更新、保存を管理
 */

import fs from 'fs';
import path from 'path';
import { validateRegistry } from '../../../validator/src/index.js';
import { getLogger } from './logger.js';

const logger = getLogger();

/**
 * RegistryManagerクラス
 */
export class RegistryManager {
  constructor(options = {}) {
    this.registryPath = options.registryPath;
    this.projectRoot = options.projectRoot || process.cwd();
    this.registry = null;
    this.validateOnSave = options.validateOnSave !== false;
  }

  /**
   * レジストリパスを解決
   */
  _resolveRegistryPath() {
    if (!this.registryPath) {
      throw new Error('レジストリパスが設定されていません');
    }

    return path.isAbsolute(this.registryPath)
      ? this.registryPath
      : path.join(this.projectRoot, this.registryPath);
  }

  /**
   * レジストリを読み込み
   */
  load() {
    const fullPath = this._resolveRegistryPath();

    if (!fs.existsSync(fullPath)) {
      throw new Error(`レジストリファイルが見つかりません: ${fullPath}`);
    }

    try {
      const content = fs.readFileSync(fullPath, 'utf-8');
      this.registry = JSON.parse(content);
      logger.debug(`レジストリ読み込み成功: ${fullPath}`);
      return this.registry;
    } catch (error) {
      throw new Error(`レジストリの読み込みに失敗しました: ${error.message}`);
    }
  }

  /**
   * レジストリを保存
   */
  save(registry = null) {
    const dataToSave = registry || this.registry;

    if (!dataToSave) {
      throw new Error('保存するレジストリデータがありません');
    }

    // バリデーション実行
    if (this.validateOnSave) {
      const errors = validateRegistry(dataToSave, {
        projectRoot: this.projectRoot,
        strict: false,
        checkContent: false,
      });

      if (errors && errors.hasErrors()) {
        throw new Error(`レジストリのバリデーションに失敗しました:\n${errors.toString()}`);
      }
    }

    // metadata.lastModified を更新
    if (!dataToSave.metadata) {
      dataToSave.metadata = {};
    }
    dataToSave.metadata.lastModified = new Date().toISOString();

    const fullPath = this._resolveRegistryPath();
    const dirPath = path.dirname(fullPath);

    // ディレクトリが存在しない場合は作成
    if (!fs.existsSync(dirPath)) {
      fs.mkdirSync(dirPath, { recursive: true });
    }

    try {
      const content = JSON.stringify(dataToSave, null, 2);
      fs.writeFileSync(fullPath, content, 'utf-8');
      this.registry = dataToSave;
      logger.debug(`レジストリ保存成功: ${fullPath}`);
      return true;
    } catch (error) {
      throw new Error(`レジストリの保存に失敗しました: ${error.message}`);
    }
  }

  /**
   * レジストリを取得
   */
  get() {
    if (!this.registry) {
      this.load();
    }
    return this.registry;
  }

  /**
   * プロジェクトを検索
   */
  findProject(projectId) {
    const registry = this.get();
    return registry.projects?.find(p => p.id === projectId);
  }

  /**
   * プロジェクトを追加
   */
  addProject(project) {
    const registry = this.get();

    if (!registry.projects) {
      registry.projects = [];
    }

    // 重複チェック
    if (this.findProject(project.id)) {
      throw new Error(`プロジェクト "${project.id}" は既に存在します`);
    }

    registry.projects.push(project);
    logger.debug(`プロジェクト追加: ${project.id}`);
    return project;
  }

  /**
   * プロジェクトを更新
   */
  updateProject(projectId, updates) {
    const registry = this.get();
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    Object.assign(project, updates);
    logger.debug(`プロジェクト更新: ${projectId}`);
    return project;
  }

  /**
   * プロジェクトを削除
   */
  removeProject(projectId) {
    const registry = this.get();
    const index = registry.projects?.findIndex(p => p.id === projectId);

    if (index === -1 || index === undefined) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    const removed = registry.projects.splice(index, 1)[0];
    logger.debug(`プロジェクト削除: ${projectId}`);
    return removed;
  }

  /**
   * ドキュメントを検索（docIdまたはslugで検索）
   */
  findDocument(projectId, identifier) {
    const project = this.findProject(projectId);
    if (!project) return null;

    return project.documents?.find(d => d.id === identifier || d.slug === identifier);
  }

  /**
   * ドキュメントを追加
   */
  addDocument(projectId, document) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    if (!project.documents) {
      project.documents = [];
    }

    // 重複チェック
    if (this.findDocument(projectId, document.id)) {
      throw new Error(`ドキュメント "${document.id}" は既に存在します`);
    }

    project.documents.push(document);
    logger.debug(`ドキュメント追加: ${projectId}/${document.id}`);
    return document;
  }

  /**
   * ドキュメントを更新（docIdまたはslugで検索）
   */
  updateDocument(projectId, identifier, updates) {
    const document = this.findDocument(projectId, identifier);

    if (!document) {
      throw new Error(`ドキュメント "${projectId}/${identifier}" が見つかりません`);
    }

    Object.assign(document, updates);
    logger.debug(`ドキュメント更新: ${projectId}/${identifier}`);
    return document;
  }

  /**
   * ドキュメントを削除（docIdまたはslugで検索）
   */
  removeDocument(projectId, identifier) {
    const project = this.findProject(projectId);
    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    // docIdまたはslugで検索
    const index = project.documents?.findIndex(d => d.id === identifier || d.slug === identifier);
    if (index === -1 || index === undefined) {
      throw new Error(`ドキュメント "${projectId}/${identifier}" が見つかりません`);
    }

    const removed = project.documents.splice(index, 1)[0];
    logger.debug(`ドキュメント削除: ${projectId}/${identifier}`);
    return removed;
  }

  /**
   * バージョンを追加
   */
  addVersion(projectId, version) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    if (!project.versions) {
      project.versions = [];
    }

    // 重複チェック
    if (project.versions.find(v => v.id === version.id)) {
      throw new Error(`バージョン "${version.id}" は既に存在します`);
    }

    project.versions.push(version);
    logger.debug(`バージョン追加: ${projectId}/${version.id}`);
    return version;
  }

  /**
   * 言語を追加
   */
  addLanguage(projectId, language) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    if (!project.languages) {
      project.languages = [];
    }

    // 重複チェック
    if (project.languages.find(l => l.code === language.code)) {
      throw new Error(`言語 "${language.code}" は既に存在します`);
    }

    project.languages.push(language);
    logger.debug(`言語追加: ${projectId}/${language.code}`);
    return language;
  }

  /**
   * バージョンを更新
   */
  updateVersion(projectId, versionId, updates) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    const version = project.versions?.find(v => v.id === versionId);
    if (!version) {
      throw new Error(`バージョン "${versionId}" が見つかりません`);
    }

    Object.assign(version, updates);
    logger.debug(`バージョン更新: ${projectId}/${versionId}`);
    return version;
  }

  /**
   * バージョンを削除
   */
  removeVersion(projectId, versionId) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    const index = project.versions?.findIndex(v => v.id === versionId);
    if (index === -1 || index === undefined) {
      throw new Error(`バージョン "${versionId}" が見つかりません`);
    }

    const removed = project.versions.splice(index, 1)[0];
    logger.debug(`バージョン削除: ${projectId}/${versionId}`);
    return removed;
  }

  /**
   * 言語を更新
   */
  updateLanguage(projectId, langCode, updates) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    const language = project.languages?.find(l => l.code === langCode);
    if (!language) {
      throw new Error(`言語 "${langCode}" が見つかりません`);
    }

    Object.assign(language, updates);
    logger.debug(`言語更新: ${projectId}/${langCode}`);
    return language;
  }

  /**
   * 言語を削除
   */
  removeLanguage(projectId, langCode) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    const index = project.languages?.findIndex(l => l.code === langCode);
    if (index === -1 || index === undefined) {
      throw new Error(`言語 "${langCode}" が見つかりません`);
    }

    const removed = project.languages.splice(index, 1)[0];
    logger.debug(`言語削除: ${projectId}/${langCode}`);
    return removed;
  }

  /**
   * 次のドキュメントIDを生成
   */
  getNextDocId(projectId) {
    const project = this.findProject(projectId);

    if (!project) {
      throw new Error(`プロジェクト "${projectId}" が見つかりません`);
    }

    if (!project.documents || project.documents.length === 0) {
      return `${projectId}-001`;
    }

    // 既存のIDから最大番号を取得
    const maxNum = project.documents.reduce((max, doc) => {
      const match = doc.id.match(/-(\d+)$/);
      if (match) {
        const num = parseInt(match[1], 10);
        return Math.max(max, num);
      }
      return max;
    }, 0);

    const nextNum = (maxNum + 1).toString().padStart(3, '0');
    return `${projectId}-${nextNum}`;
  }

  /**
   * レジストリ全体のバリデーション
   */
  validate(options = {}) {
    const registry = this.get();
    return validateRegistry(registry, {
      projectRoot: this.projectRoot,
      ...options,
    });
  }
}

/**
 * レジストリマネージャーを作成
 */
export function createRegistryManager(options) {
  return new RegistryManager(options);
}

/**
 * エクスポート
 */
export default RegistryManager;
