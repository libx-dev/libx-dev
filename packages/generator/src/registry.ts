/**
 * レジストリローダー
 *
 * registry/docs.jsonを読み込み、バリデーションを実行して返します。
 */

import { readFileSync } from 'fs';
import { resolve } from 'path';
import type { Registry } from './types.js';

/**
 * レジストリファイルのデフォルトパス
 */
const DEFAULT_REGISTRY_PATH = 'registry/docs.json';

/**
 * レジストリローダーのエラークラス
 */
export class RegistryLoadError extends Error {
  constructor(
    message: string,
    public readonly cause?: Error
  ) {
    super(message);
    this.name = 'RegistryLoadError';
  }
}

/**
 * レジストリを読み込み
 *
 * @param registryPath - レジストリファイルのパス（デフォルト: registry/docs.json）
 * @param basePath - ベースディレクトリ（デフォルト: process.cwd()）
 * @returns パース済みのレジストリオブジェクト
 * @throws {RegistryLoadError} ファイル読み込みまたはパースに失敗した場合
 *
 * @example
 * ```ts
 * const registry = loadRegistry();
 * console.log(registry.projects.length);
 * ```
 */
export function loadRegistry(
  registryPath: string = DEFAULT_REGISTRY_PATH,
  basePath: string = process.cwd()
): Registry {
  const fullPath = resolve(basePath, registryPath);

  try {
    // ファイル読み込み
    const content = readFileSync(fullPath, 'utf-8');

    // JSONパース
    const registry = JSON.parse(content) as Registry;

    // 基本構造の検証
    validateRegistryStructure(registry);

    return registry;
  } catch (error) {
    if (error instanceof RegistryLoadError) {
      throw error;
    }

    if (error instanceof SyntaxError) {
      throw new RegistryLoadError(
        `Failed to parse registry JSON at ${fullPath}: ${error.message}`,
        error
      );
    }

    if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
      throw new RegistryLoadError(
        `Registry file not found at ${fullPath}`,
        error as Error
      );
    }

    throw new RegistryLoadError(
      `Failed to load registry from ${fullPath}: ${error}`,
      error as Error
    );
  }
}

/**
 * レジストリの基本構造を検証
 *
 * @param registry - 検証するレジストリオブジェクト
 * @throws {RegistryLoadError} 構造が不正な場合
 */
function validateRegistryStructure(registry: any): asserts registry is Registry {
  if (!registry || typeof registry !== 'object') {
    throw new RegistryLoadError('Registry must be an object');
  }

  if (!registry.$schemaVersion) {
    throw new RegistryLoadError('Registry must have $schemaVersion field');
  }

  if (!registry.metadata || typeof registry.metadata !== 'object') {
    throw new RegistryLoadError('Registry must have metadata field');
  }

  if (!Array.isArray(registry.projects)) {
    throw new RegistryLoadError('Registry must have projects array');
  }

  if (registry.projects.length === 0) {
    throw new RegistryLoadError('Registry must have at least one project');
  }

  // プロジェクトの基本検証
  registry.projects.forEach((project: any, index: number) => {
    if (!project.id || typeof project.id !== 'string') {
      throw new RegistryLoadError(
        `Project at index ${index} must have a valid id field`
      );
    }

    if (!Array.isArray(project.documents)) {
      throw new RegistryLoadError(
        `Project "${project.id}" must have documents array`
      );
    }

    if (!Array.isArray(project.languages)) {
      throw new RegistryLoadError(
        `Project "${project.id}" must have languages array`
      );
    }

    if (!Array.isArray(project.versions)) {
      throw new RegistryLoadError(
        `Project "${project.id}" must have versions array`
      );
    }
  });
}

/**
 * レジストリから特定のプロジェクトを取得
 *
 * @param registry - レジストリオブジェクト
 * @param projectId - プロジェクトID
 * @returns プロジェクトオブジェクト、見つからない場合はundefined
 */
export function getProject(registry: Registry, projectId: string) {
  return registry.projects.find((p) => p.id === projectId);
}

/**
 * レジストリから特定のドキュメントを取得
 *
 * @param registry - レジストリオブジェクト
 * @param projectId - プロジェクトID
 * @param docId - ドキュメントID
 * @returns ドキュメントオブジェクト、見つからない場合はundefined
 */
export function getDocument(registry: Registry, projectId: string, docId: string) {
  const project = getProject(registry, projectId);
  if (!project) return undefined;

  return project.documents.find((d) => d.id === docId);
}

/**
 * プロジェクトの統計情報を取得
 *
 * @param registry - レジストリオブジェクト
 * @returns 統計情報オブジェクト
 */
export function getRegistryStats(registry: Registry) {
  const stats = {
    projectCount: registry.projects.length,
    totalDocuments: 0,
    totalLanguages: new Set<string>(),
    totalVersions: new Set<string>(),
    documentsByProject: {} as Record<string, number>,
  };

  registry.projects.forEach((project) => {
    stats.totalDocuments += project.documents.length;
    stats.documentsByProject[project.id] = project.documents.length;

    project.languages.forEach((lang) => {
      stats.totalLanguages.add(lang.code);
    });

    project.versions.forEach((version) => {
      stats.totalVersions.add(version.id);
    });
  });

  return {
    ...stats,
    totalLanguages: stats.totalLanguages.size,
    totalVersions: stats.totalVersions.size,
  };
}
