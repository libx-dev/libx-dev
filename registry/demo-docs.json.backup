{
  "$schemaVersion": "1.0.0",
  "metadata": {
    "createdAt": "2025-10-20T00:00:00.000Z",
    "createdBy": "phase-2-6-demo-project"
  },
  "projects": [
    {
      "id": "demo-docs",
      "displayName": {
        "en": "Demo Documentation",
        "ja": "デモドキュメント",
        "ko": "데모 문서"
      },
      "description": {
        "en": "Demonstration of the new documentation generator with shared packages",
        "ja": "共有パッケージを活用した新ドキュメントジェネレーターのデモ",
        "ko": "공유 패키지를 활용한 새 문서 생성기 데모"
      },
      "languages": [
        {
          "code": "en",
          "displayName": "English",
          "status": "active",
          "default": true
        },
        {
          "code": "ja",
          "displayName": "日本語",
          "status": "active",
          "fallback": "en"
        },
        {
          "code": "ko",
          "displayName": "한국어",
          "status": "active",
          "fallback": "en"
        }
      ],
      "versions": [
        {
          "id": "v1",
          "name": "Version 1.0",
          "isLatest": true,
          "status": "stable",
          "date": "2025-10-20T00:00:00.000Z"
        }
      ],
      "categories": [
        {
          "id": "guide",
          "name": {
            "en": "Guide",
            "ja": "ガイド",
            "ko": "가이드"
          },
          "order": 1,
          "description": {
            "en": "Getting started and user guides",
            "ja": "はじめに、ユーザーガイド",
            "ko": "시작하기 및 사용자 가이드"
          }
        },
        {
          "id": "api",
          "name": {
            "en": "API Reference",
            "ja": "APIリファレンス",
            "ko": "API 참조"
          },
          "order": 2,
          "description": {
            "en": "API documentation and reference",
            "ja": "APIドキュメントとリファレンス",
            "ko": "API 문서 및 참조"
          }
        },
        {
          "id": "examples",
          "name": {
            "en": "Examples",
            "ja": "サンプル",
            "ko": "예제"
          },
          "order": 3,
          "description": {
            "en": "Code examples and use cases",
            "ja": "コード例とユースケース",
            "ko": "코드 예제 및 사용 사례"
          }
        }
      ],
      "documents": [
        {
          "docId": "getting-started",
          "projectId": "demo-docs",
          "versionId": "v1",
          "categoryId": "guide",
          "order": 1,
          "slug": "getting-started",
          "title": {
            "en": "Getting Started",
            "ja": "はじめに",
            "ko": "시작하기"
          },
          "description": {
            "en": "Quick start guide for the new documentation generator",
            "ja": "新ドキュメントジェネレーターのクイックスタートガイド",
            "ko": "새 문서 생성기 빠른 시작 가이드"
          },
          "visibility": "public"
        },
        {
          "docId": "installation",
          "projectId": "demo-docs",
          "versionId": "v1",
          "categoryId": "guide",
          "order": 2,
          "slug": "installation",
          "title": {
            "en": "Installation",
            "ja": "インストール",
            "ko": "설치"
          },
          "description": {
            "en": "Installation guide and requirements",
            "ja": "インストールガイドと要件",
            "ko": "설치 가이드 및 요구 사항"
          },
          "visibility": "public"
        },
        {
          "docId": "configuration",
          "projectId": "demo-docs",
          "versionId": "v1",
          "categoryId": "guide",
          "order": 3,
          "slug": "configuration",
          "title": {
            "en": "Configuration",
            "ja": "設定",
            "ko": "구성"
          },
          "description": {
            "en": "Configuration options and customization",
            "ja": "設定オプションとカスタマイズ",
            "ko": "구성 옵션 및 사용자 정의"
          },
          "visibility": "public"
        },
        {
          "docId": "components",
          "projectId": "demo-docs",
          "versionId": "v1",
          "categoryId": "api",
          "order": 1,
          "slug": "components",
          "title": {
            "en": "Components",
            "ja": "コンポーネント",
            "ko": "컴포넌트"
          },
          "description": {
            "en": "UI components API reference",
            "ja": "UIコンポーネントAPIリファレンス",
            "ko": "UI 컴포넌트 API 참조"
          },
          "visibility": "public"
        },
        {
          "docId": "utilities",
          "projectId": "demo-docs",
          "versionId": "v1",
          "categoryId": "api",
          "order": 2,
          "slug": "utilities",
          "title": {
            "en": "Utilities",
            "ja": "ユーティリティ",
            "ko": "유틸리티"
          },
          "description": {
            "en": "Utility functions and helpers",
            "ja": "ユーティリティ関数とヘルパー",
            "ko": "유틸리티 함수 및 헬퍼"
          },
          "visibility": "public"
        },
        {
          "docId": "basic-usage",
          "projectId": "demo-docs",
          "versionId": "v1",
          "categoryId": "examples",
          "order": 1,
          "slug": "basic-usage",
          "title": {
            "en": "Basic Usage",
            "ja": "基本的な使用方法",
            "ko": "기본 사용법"
          },
          "description": {
            "en": "Basic usage examples",
            "ja": "基本的な使用例",
            "ko": "기본 사용 예제"
          },
          "visibility": "public"
        }
      ]
    }
  ],
  "glossaryTerms": [
    {
      "termId": "registry",
      "term": {
        "en": "Registry",
        "ja": "レジストリ",
        "ko": "레지스트리"
      },
      "definition": {
        "en": "A JSON-based data structure that defines the documentation structure, including projects, versions, categories, and documents.",
        "ja": "プロジェクト、バージョン、カテゴリ、ドキュメントを定義するJSONベースのデータ構造です。",
        "ko": "프로젝트, 버전, 카테고리, 문서를 정의하는 JSON 기반 데이터 구조입니다."
      },
      "relatedDocs": ["getting-started", "configuration"]
    },
    {
      "termId": "shared-packages",
      "term": {
        "en": "Shared Packages",
        "ja": "共有パッケージ",
        "ko": "공유 패키지"
      },
      "definition": {
        "en": "Reusable packages including @docs/generator, @docs/ui, @docs/theme, @docs/i18n, and @docs/versioning that provide core functionality.",
        "ja": "@docs/generator、@docs/ui、@docs/theme、@docs/i18n、@docs/versioningを含む、コア機能を提供する再利用可能なパッケージです。",
        "ko": "핵심 기능을 제공하는 @docs/generator, @docs/ui, @docs/theme, @docs/i18n, @docs/versioning을 포함한 재사용 가능한 패키지입니다."
      },
      "relatedDocs": ["installation", "components"]
    }
  ]
}
