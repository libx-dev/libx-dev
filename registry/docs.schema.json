{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "docs.schema.json",
  "title": "LibX Docs Registry Schema",
  "description": "新ドキュメントジェネレーター用のレジストリ定義。CLI・ビルド・移行ツールの単一ソースを提供する。",
  "type": "object",
  "required": [
    "$schemaVersion",
    "projects",
    "settings"
  ],
  "additionalProperties": false,
  "properties": {
    "$schemaVersion": {
      "description": "スキーマの SemVer。互換性判定とマイグレーション判定に利用する。",
      "type": "string",
      "pattern": "^\\d+\\.\\d+\\.\\d+$",
      "examples": [
        "1.0.0",
        "1.1.0"
      ]
    },
    "metadata": {
      "description": "レジストリ生成日時や担当者などの運用メタデータ。キーは任意。",
      "type": "object",
      "additionalProperties": true
    },
    "projects": {
      "description": "管理対象プロジェクトのリスト。`id` は一意である必要がある。",
      "type": "array",
      "minItems": 1,
      "items": {
        "$ref": "./schema/project.schema.json"
      }
    },
    "settings": {
      "description": "レジストリ全体に適用される設定値。",
      "$ref": "./schema/settings.schema.json"
    }
  }
}
