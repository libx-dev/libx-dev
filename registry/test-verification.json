{
  "$schemaVersion": "1.0.0",
  "metadata": {
    "createdAt": "2025-10-16T05:07:31.187Z",
    "createdBy": "migrate-to-registry.js"
  },
  "projects": [
    {
      "id": "test-verification",
      "displayName": {
        "en": "Test Verification",
        "ja": "検証テスト",
        "ko": "검증 테스트"
      },
      "description": {
        "en": "Documentation for Test Verification",
        "ja": "検証テストのドキュメントです",
        "ko": "검증 테스트 문서입니다"
      },
      "languages": [
        {
          "code": "en",
          "displayName": "English",
          "status": "active",
          "default": true
        },
        {
          "code": "ja",
          "displayName": "日本語",
          "status": "active",
          "fallback": "en"
        },
        {
          "code": "ko",
          "displayName": "한국어",
          "status": "active",
          "fallback": "en"
        }
      ],
      "versions": [
        {
          "id": "v1",
          "name": "Version 1.0",
          "isLatest": false,
          "status": "deprecated",
          "date": "2024-01-01T00:00:00.000Z"
        },
        {
          "id": "v2",
          "name": "Version 2",
          "isLatest": true,
          "status": "active",
          "date": "2025-08-30T02:13:09.787Z"
        }
      ],
      "categories": [
        {
          "id": "guide",
          "order": 1,
          "titles": {
            "en": "Guide",
            "ja": "ガイド",
            "ko": "가이드"
          },
          "docs": [
            "installation-guide",
            "getting-started"
          ]
        }
      ],
      "documents": [
        {
          "id": "api-overview",
          "slug": "api-reference/api-overview",
          "title": {
            "ko": "API 개요",
            "ja": "API概要",
            "en": "API Overview"
          },
          "summary": {
            "ko": "API 개요에 대한 설명",
            "ja": "API概要について説明します",
            "en": "API Overviewについて説明します"
          },
          "versions": [
            "v2"
          ],
          "status": "published",
          "visibility": "public",
          "keywords": [],
          "tags": [],
          "content": {
            "ko": {
              "path": "content/api-overview/ko.mdx",
              "status": "published"
            },
            "ja": {
              "path": "content/api-overview/ja.mdx",
              "status": "published"
            },
            "en": {
              "path": "content/api-overview/en.mdx",
              "status": "published"
            }
          },
          "license": "test-verification-original"
        },
        {
          "id": "installation-guide",
          "slug": "guide/installation-guide",
          "title": {
            "ko": "설치 가이드",
            "ja": "インストールガイド",
            "en": "Installation Guide"
          },
          "summary": {
            "ko": "검증 테스트 v2 설치 방법",
            "ja": "インストールガイドについて説明します",
            "en": "Installation Guideについて説明します"
          },
          "versions": [
            "v2"
          ],
          "status": "published",
          "visibility": "public",
          "keywords": [],
          "tags": [],
          "content": {
            "ko": {
              "path": "content/installation-guide/ko.mdx",
              "status": "published"
            },
            "ja": {
              "path": "content/installation-guide/ja.mdx",
              "status": "published"
            },
            "en": {
              "path": "content/installation-guide/en.mdx",
              "status": "published"
            }
          },
          "license": "test-verification-original"
        },
        {
          "id": "getting-started",
          "slug": "guide/getting-started",
          "title": {
            "ko": "검증 테스트에 오신 것을 환영합니다",
            "ja": "プロジェクトテンプレートへようこそ",
            "en": "Welcome to Project Template"
          },
          "summary": {
            "ko": "새로운 문서 사이트 생성을 위한 시작점",
            "ja": "新しいドキュメントサイト作成の出発点",
            "en": "Starting point for creating new documentation sites"
          },
          "versions": [
            "v2",
            "v1"
          ],
          "status": "published",
          "visibility": "public",
          "keywords": [],
          "tags": [],
          "content": {
            "ko": {
              "path": "content/getting-started/ko.mdx",
              "status": "published"
            },
            "ja": {
              "path": "content/getting-started/ja.mdx",
              "status": "published"
            },
            "en": {
              "path": "content/getting-started/en.mdx",
              "status": "published"
            }
          },
          "license": "test-verification-original"
        }
      ],
      "licenses": [
        {
          "id": "test-verification-original",
          "name": "MIT",
          "url": "https://opensource.org/licenses/MIT",
          "attribution": "Test Verification Documentation by libx-dev Team (https://github.com/libx-dev/libx-dev)"
        }
      ],
      "visibility": "public",
      "options": {
        "search": true
      }
    }
  ],
  "settings": {
    "defaultLocale": "en",
    "defaultVersion": "v1",
    "build": {
      "search": "pagefind",
      "sitemap": true
    },
    "deploy": {
      "provider": "cloudflare-pages",
      "basePath": "/"
    },
    "backup": {
      "enabled": true,
      "retentionDays": 14
    }
  }
}