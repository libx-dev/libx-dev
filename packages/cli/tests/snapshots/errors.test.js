/**
 * エラーメッセージのスナップショットテスト
 *
 * エラーメッセージの一貫性を保証するためのテスト
 */

import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { ValidationErrorCollection, ValidationError, ValidationWarning } from '../../../validator/src/index.js';
import { RegistryManager } from '../../src/utils/registry.js';
import { setupTest } from '../helpers/fixtures.js';

describe('エラーメッセージスナップショット', () => {
  let testEnv;

  beforeEach(() => {
    testEnv = setupTest('snapshot');
  });

  afterEach(() => {
    testEnv.cleanup();
  });

  describe('RegistryManagerエラー', () => {
    it('存在しないプロジェクトエラー', () => {
      const manager = new RegistryManager({
        registryPath: 'registry/docs.json',
        projectRoot: testEnv.tempDir,
      });
      manager.load();

      try {
        manager.findProject('nonexistent');
        manager.updateProject('nonexistent', { displayName: { en: 'Test' } });
      } catch (error) {
        expect(error.message).toMatchSnapshot();
      }
    });

    it('重複プロジェクトエラー', () => {
      const manager = new RegistryManager({
        registryPath: 'registry/docs.json',
        projectRoot: testEnv.tempDir,
        validateOnSave: false,
      });
      manager.load();

      try {
        manager.addProject({
          id: 'test-project-1',
          displayName: { en: 'Duplicate' },
        });
      } catch (error) {
        expect(error.message).toMatchSnapshot();
      }
    });

    it('存在しないドキュメントエラー', () => {
      const manager = new RegistryManager({
        registryPath: 'registry/docs.json',
        projectRoot: testEnv.tempDir,
      });
      manager.load();

      try {
        manager.updateDocument('test-project-1', 'nonexistent', { title: { en: 'Test' } });
      } catch (error) {
        expect(error.message).toMatchSnapshot();
      }
    });
  });

  describe('ValidationErrorCollection', () => {
    it('単一エラーの文字列表現', () => {
      const collection = new ValidationErrorCollection();
      const error = new ValidationError(
        'MISSING_FIELD',
        'Required field "displayName" is missing',
        null,
        'projects[0].displayName'
      );
      collection.add(error);

      expect(collection.toString()).toMatchSnapshot();
    });

    it('複数エラーと警告の文字列表現', () => {
      const collection = new ValidationErrorCollection();
      collection.add(new ValidationError('MISSING_FIELD', 'Required field "displayName" is missing'));
      collection.add(new ValidationError('INVALID_ID', 'Project ID contains invalid characters'));
      collection.add(new ValidationWarning('DEPRECATED_FIELD', 'Field "oldField" is deprecated'));

      expect(collection.toString()).toMatchSnapshot();
    });

    it('エラー件数サマリー', () => {
      const collection = new ValidationErrorCollection();
      collection.add(new ValidationError('ERROR_1', 'Error 1'));
      collection.add(new ValidationError('ERROR_2', 'Error 2'));
      collection.add(new ValidationWarning('WARNING_1', 'Warning 1'));

      const summary = collection.getSummary();
      expect(summary).toMatchSnapshot();
    });
  });

  describe('Logger出力形式', () => {
    it('JSONモード出力形式', () => {
      const logEntry = {
        level: 'error',
        message: 'Failed to load registry',
        data: {
          path: 'registry/docs.json',
          error: 'File not found',
        },
        timestamp: '2025-01-01T00:00:00.000Z',
      };

      expect(JSON.stringify(logEntry, null, 2)).toMatchSnapshot();
    });
  });
});
