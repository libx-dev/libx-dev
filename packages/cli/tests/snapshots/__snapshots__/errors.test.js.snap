// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`エラーメッセージスナップショット > Logger出力形式 > JSONモード出力形式 1`] = `
"{
  "level": "error",
  "message": "Failed to load registry",
  "data": {
    "path": "registry/docs.json",
    "error": "File not found"
  },
  "timestamp": "2025-01-01T00:00:00.000Z"
}"
`;

exports[`エラーメッセージスナップショット > RegistryManagerエラー > 存在しないドキュメントエラー 1`] = `"ドキュメント "test-project-1/nonexistent" が見つかりません"`;

exports[`エラーメッセージスナップショット > RegistryManagerエラー > 存在しないプロジェクトエラー 1`] = `"プロジェクト "nonexistent" が見つかりません"`;

exports[`エラーメッセージスナップショット > RegistryManagerエラー > 重複プロジェクトエラー 1`] = `"プロジェクト "test-project-1" は既に存在します"`;

exports[`エラーメッセージスナップショット > ValidationErrorCollection > エラー件数サマリー 1`] = `
{
  "errorCount": 2,
  "totalCount": 3,
  "warningCount": 1,
}
`;

exports[`エラーメッセージスナップショット > ValidationErrorCollection > 単一エラーの文字列表現 1`] = `
"[31m✗[0m [MISSING_FIELD] Required field "displayName" is missing
  [90m位置:[0m projects[0].displayName"
`;

exports[`エラーメッセージスナップショット > ValidationErrorCollection > 複数エラーと警告の文字列表現 1`] = `
"[31m✗[0m [MISSING_FIELD] Required field "displayName" is missing

[31m✗[0m [INVALID_ID] Project ID contains invalid characters

[33m⚠[0m [DEPRECATED_FIELD] Field "oldField" is deprecated"
`;
